File: examples/desktop-automation/browser_automation.py
"""
浏览器自动化示例任务
功能：打开网页，执行搜索，截图保存
依赖：pip install selenium playwright
"""
import asyncio
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time
import os
import json
from pathlib import Path
from autoflow_sdk import TaskContext


def get_typed_param(ctx, key, default=None):
    """
    JSON 容错解析任务参数。

    执行器把所有触发参数字符串化注入（AUTOFLOW_* 环境变量），python SDK
    from_env 不做类型还原——列表/数字/布尔参数拿到的都是字符串。尝试
    json.loads 还原，失败则原样返回字符串（对齐 Node 示例 getParam 的兜底）。
    """
    raw = ctx.get_param(key)
    if raw is None:
        return default
    try:
        return json.loads(raw)
    except (ValueError, TypeError):
        return raw


def is_truthy_param(value) -> bool:
    """布尔参数显式判定：字符串 "true"/"1" 为真，"false"/"0" 为假。
    避免 if value: 对非空字符串（如 "false"）恒真的真值反转。"""
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        return value.strip().lower() in ("true", "1", "yes")
    return bool(value)


def main():
    ctx = TaskContext.from_env()
    
    # 获取任务参数（JSON 容错解析 + 布尔显式比较，见 helper 注释）
    url = get_typed_param(ctx, "url", "https://www.baidu.com")
    search_keyword = get_typed_param(ctx, "keyword", "AutoCodeFlow")
    headless = is_truthy_param(get_typed_param(ctx, "headless", False))
    
    ctx.log.info(f"开始浏览器自动化任务")
    ctx.log.info(f"目标URL: {url}")
    ctx.log.info(f"搜索关键词: {search_keyword}")
    
    # 创建输出目录
    output_dir = Path(f"/tmp/browser_automation_{ctx.execution_id}")
    output_dir.mkdir(parents=True, exist_ok=True)
    
    try:
        # 启动浏览器
        options = webdriver.ChromeOptions()
        if headless:
            options.add_argument('--headless')
        options.add_argument('--no-sandbox')
        options.add_argument('--disable-dev-shm-usage')
        options.add_argument('--window-size=1920,1080')
        
        driver = webdriver.Chrome(options=options)
        driver.set_window_size(1920, 1080)
        
        ctx.log.info("浏览器启动成功")
        
        # 打开网页
        driver.get(url)
        ctx.log.info(f"已打开: {url}")
        time.sleep(2)
        
        # 截图 - 初始页面
        initial_screenshot = output_dir / "01_initial_page.png"
        driver.save_screenshot(str(initial_screenshot))
        ctx.log.info(f"已保存初始截图: {initial_screenshot}")
        
        # 执行搜索
        search_box = WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.ID, "kw"))
        )
        search_box.clear()
        search_box.send_keys(search_keyword)
        
        search_button = driver.find_element(By.ID, "su")
        search_button.click()
        
        ctx.log.info(f"已执行搜索: {search_keyword}")
        time.sleep(3)
        
        # 截图 - 搜索结果
        results_screenshot = output_dir / "02_search_results.png"
        driver.save_screenshot(str(results_screenshot))
        ctx.log.info(f"已保存搜索结果截图: {results_screenshot}")
        
        # 获取搜索结果
        results = driver.find_elements(By.CSS_SELECTOR, ".result h3 a")
        ctx.log.info(f"找到 {len(results)} 个搜索结果")
        
        # 保存搜索结果到文件
        results_file = output_dir / "search_results.txt"
        with open(results_file, 'w', encoding='utf-8') as f:
            for i, result in enumerate(results[:10], 1):
                try:
                    text = result.text
                    href = result.get_attribute('href')
                    f.write(f"{i}. {text}\n")
                    f.write(f"   链接: {href}\n\n")
                except Exception as e:
                    ctx.log.warning(f"获取第 {i} 个结果失败: {e}")
        
        ctx.log.info(f"搜索结果已保存到: {results_file}")
        
        # 点击第一个结果
        if results:
            first_result = results[0]
            first_result.click()
            ctx.log.info("已点击第一个搜索结果")
            time.sleep(3)
            
            # 截图 - 目标页面
            target_screenshot = output_dir / "03_target_page.png"
            driver.save_screenshot(str(target_screenshot))
            ctx.log.info(f"已保存目标页面截图: {target_screenshot}")
        
        # 关闭浏览器
        driver.quit()
        ctx.log.info("浏览器已关闭")
        
        # 返回结果
        result = {
            "success": True,
            "message": "浏览器自动化任务完成",
            "output_dir": str(output_dir),
            "screenshots": [
                str(initial_screenshot),
                str(results_screenshot),
                str(target_screenshot)
            ],
            "results_file": str(results_file),
            "search_results_count": len(results)
        }
        
        print(f"RESULT: {result}")
        return result
        
    except Exception as e:
        ctx.log.error(f"任务执行失败: {str(e)}")
        error_result = {
            "success": False,
            "error": str(e),
            "output_dir": str(output_dir)
        }
        print(f"RESULT: {error_result}")
        return error_result


if __name__ == "__main__":
    main()

File: examples/desktop-automation/browser_automation.js
/**
 * 浏览器自动化示例任务 (Node.js版本)
 * 功能：打开网页，执行搜索，截图保存
 * 依赖：npm install puppeteer
 */

const puppeteer = require('puppeteer');
const path = require('path');
const fs = require('fs').promises;

// 从环境变量获取上下文（在实际执行时由执行器注入）
const executionId = process.env.EXECUTION_ID || 'local-test';
const taskId = process.env.TASK_ID || 'browser-auto-demo';
const taskName = process.env.TASK_NAME || '浏览器自动化演示';

// 简单的日志记录器
const logger = {
  info: (msg) => console.log(`[INFO] ${msg}`),
  error: (msg) => console.error(`[ERROR] ${msg}`),
  warning: (msg) => console.warn(`[WARNING] ${msg}`)
};

// 从环境变量或默认值获取参数
const getParam = (key, defaultValue) => {
  const envKey = `AUTOFLOW_${key.toUpperCase()}`;
  return process.env[envKey] !== undefined ? JSON.parse(process.env[envKey]) : defaultValue;
};

async function main() {
  logger.info(`开始浏览器自动化任务 (执行ID: ${executionId})`);
  
  // 获取任务参数
  const url = getParam('url', 'https://www.baidu.com');
  const searchKeyword = getParam('keyword', 'AutoCodeFlow');
  const headless = getParam('headless', false);
  
  logger.info(`目标URL: ${url}`);
  logger.info(`搜索关键词: ${searchKeyword}`);
  logger.info(`无头模式: ${headless}`);
  
  // 创建输出目录
  const outputDir = path.join('/tmp', `browser_automation_${executionId}`);
  await fs.mkdir(outputDir, { recursive: true });
  
  let browser;
  try {
    // 启动浏览器
    browser = await puppeteer.launch({
      headless: headless,
      args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--window-size=1920,1080'
      ]
    });
    
    const page = await browser.newPage();
    await page.setViewport({ width: 1920, height: 1080 });
    
    logger.info('浏览器启动成功');
    
    // 打开网页
    await page.goto(url, { waitUntil: 'networkidle2' });
    logger.info(`已打开: ${url}`);
    
    // 截图 - 初始页面
    const initialScreenshot = path.join(outputDir, '01_initial_page.png');
    await page.screenshot({ path: initialScreenshot, fullPage: true });
    logger.info(`已保存初始截图: ${initialScreenshot}`);
    
    // 执行搜索
    await page.type('#kw', searchKeyword);
    await page.click('#su');
    
    logger.info(`已执行搜索: ${searchKeyword}`);
    
    // 等待搜索结果加载
    await page.waitForSelector('.result', { timeout: 10000 });
    await page.waitForTimeout(2000);
    
    // 截图 - 搜索结果
    const resultsScreenshot = path.join(outputDir, '02_search_results.png');
    await page.screenshot({ path: resultsScreenshot, fullPage: true });
    logger.info(`已保存搜索结果截图: ${resultsScreenshot}`);
    
    // 获取搜索结果
    const searchResults = await page.evaluate(() => {
      const results = [];
      const items = document.querySelectorAll('.result h3 a');
      items.forEach((item, index) => {
        results.push({
          index: index + 1,
          text: item.textContent,
          href: item.href
        });
      });
      return results;
    });
    
    logger.info(`找到 ${searchResults.length} 个搜索结果`);
    
    // 保存搜索结果到文件
    const resultsFile = path.join(outputDir, 'search_results.txt');
    let resultsText = `搜索结果 - ${new Date().toLocaleString('zh-CN')}\n`;
    resultsText += `关键词: ${searchKeyword}\n`;
    resultsText += `${'='.repeat(50)}\n\n`;
    
    searchResults.slice(0, 10).forEach(result => {
      resultsText += `${result.index}. ${result.text}\n`;
      resultsText += `   链接: ${result.href}\n\n`;
    });
    
    await fs.writeFile(resultsFile, resultsText, 'utf-8');
    logger.info(`搜索结果已保存到: ${resultsFile}`);
    
    // 点击第一个结果
    if (searchResults.length > 0) {
      await Promise.all([
        page.waitForNavigation({ waitUntil: 'networkidle2' }),
        page.click('.result h3 a')
      ]);
      
      logger.info('已点击第一个搜索结果');
      await page.waitForTimeout(3000);
      
      // 截图 - 目标页面
      const targetScreenshot = path.join(outputDir, '03_target_page.png');
      await page.screenshot({ path: targetScreenshot, fullPage: true });
      logger.info(`已保存目标页面截图: ${targetScreenshot}`);
    }
    
    // 关闭浏览器
    await browser.close();
    logger.info('浏览器已关闭');
    
    // 返回结果
    const result = {
      success: true,
      message: '浏览器自动化任务完成',
      executionId,
      outputDir,
      screenshots: [
        initialScreenshot,
        resultsScreenshot,
        path.join(outputDir, '03_target_page.png')
      ].filter(Boolean),
      resultsFile,
      searchResultsCount: searchResults.length,
      timestamp: new Date().toISOString()
    };
    
    console.log(`RESULT: ${JSON.stringify(result, null, 2)}`);
    return result;
    
  } catch (error) {
    logger.error(`任务执行失败: ${error.message}`);
    
    if (browser) {
      await browser.close();
    }
    
    const errorResult = {
      success: false,
      error: error.message,
      executionId,
      outputDir,
      timestamp: new Date().toISOString()
    };
    
    console.log(`RESULT: ${JSON.stringify(errorResult, null, 2)}`);
    return errorResult;
  }
}

// 如果直接运行此文件
if (require.main === module) {
  main().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

module.exports = { main };